core/drivers/aplic_msi.c

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2025 Beijing Institute of Open Source Chip (BOSC)
+ */
+
+#include <assert.h>
+#include <config.h>
+#include <drivers/aplic.h>
+#include <drivers/aplic_priv.h>
+#include <io.h>
+#include <kernel/interrupt.h>
+#include <kernel/misc.h>
+#include <kernel/panic.h>
+#include <mm/core_memprot.h>
+#include <mm/core_mmu.h>
+#include <platform_config.h>
+#include <stdlib.h>
+#include <sys/queue.h>
+#include <trace.h>
+#include <types_ext.h>
+#include <util.h>
+
+#define APLIC_DEFAULT_EIID		2
+
+static struct aplic_data aplic_data __nex_bss;
+
+static void aplic_set_target(struct aplic_data *aplic, uint32_t source,
+			     uint32_t hart_idx, uint32_t guest_idx,
+			     uint32_t eiid)
+{
+	vaddr_t target = 0;
+	uint32_t val = 0;
+
+	val = SHIFT_U32(hart_idx & APLIC_TARGET_HART_IDX_MASK,
+			APLIC_TARGET_HART_IDX_SHIFT);
```

Comment:
According to `APLIC_TARGET_HART_IDX_MASK`, the mask operation should be done after shift.

```c
	val = SHIFT_U32(hart_idx, APLIC_TARGET_HART_IDX_SHIFT) & APLIC_TARGET_HART_IDX_MASK;
```